src/annotator.cpp

```diff
@@ -0,0 +1,361 @@
+/*
+Copyright libCellML Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "libcellml/annotator.h"
+
+#include <algorithm>
+
+#include "libcellml/component.h"
+#include "libcellml/importsource.h"
+#include "libcellml/model.h"
+#include "libcellml/namedentity.h"
+#include "libcellml/printer.h"
+#include "libcellml/reset.h"
+#include "libcellml/types.h"
+#include "libcellml/units.h"
+#include "libcellml/variable.h"
+
+#include "internaltypes.h"
+#include "namespaces.h"
+#include "utilities.h"
+
+namespace libcellml {
+
+struct Annotator::AnnotatorImpl
+{
+    Annotator *mAnnotator = nullptr;
+    ItemList mIdList;
+    bool isBuilt = false;
+    ModelPtr mModel = nullptr;
+};
+
+Annotator::Annotator()
+    : mPimpl(new AnnotatorImpl())
+{
+    mPimpl->mAnnotator = this;
+    mPimpl->mIdList = std::map<std::string, AnyItem>();
+    mPimpl->isBuilt = false;
+}
+
+Annotator::~Annotator()
+{
+    delete mPimpl;
+}
+
+AnnotatorPtr Annotator::create() noexcept
+{
+    return std::shared_ptr<Annotator> {new Annotator {}};
+}
+
+void listComponentIdsAndItems(const ComponentPtr &component, ItemList &idList)
+{
+    std::string id = component->id();
+    if (!id.empty()) {
+        idList.insert(std::make_pair(id, std::make_pair(0, component)));
```

Comment:
probably missing something, but shouldn't that `0` be `Annotator::Type::COMPONENT` ?

Reply:
Yes, weirdness.  Thanks!

Reply:
83ce6a11